tools/upload_text.php

```diff
@@ -249,164 +217,138 @@
     // make reasonably sure script does not timeout on large file uploads
     set_time_limit(14400);
 
-    // if files have been uploaded, process them and mangle the postcomments
-    $uploaded_file = validate_uploaded_file();
+    slim_header($title);
+    echo "<h1>$title</h1>";
+    echo "<h2>", sprintf("Project: %s", $project->nameofwork), "</h2>";
 
-    $have_file = FALSE;
-    if (is_file($uploaded_file))
+    // if files have been uploaded, process them and mangle the postcomments
+    $temporary_path = "";
+    $returning_to_pool = ('return_1' == $stage || 'return_2' == $stage);
+    try
     {
-        // replace filename
-        $zipext = ".zip";
-        $name = $projectid.$indicator.$zipext;
-        $location = "$project->dir/$name";
-        ensure_path_is_unused( $location );
-        copy($uploaded_file, $location);
-        unlink($uploaded_file);
-        $have_file = TRUE;
-        if ($stage == 'smooth_avail')
+        try
         {
-            $project->delete_smoothreading_dir();
-            $smooth_dir = "$project->dir/smooth";
-            if(!mkdir($smooth_dir))
+            // inner try succeeds only if there is a file
+            $file_info = validate_uploaded_file();
+            $have_file = true;
+            $temporary_path = $file_info["tmp_name"];
+            $original_name = $file_info['name'];
+
+            zip_check($original_name, $temporary_path);
+            // replace filename
+            $zipext = ".zip";
+            $name = $projectid.$indicator.$zipext;
+            $location = "$project->dir/$name";
+            ensure_path_is_unused( $location );
+            rename($temporary_path, $location);
+            if ($stage == 'smooth_avail')
             {
-                die("Could not create smooth directory");
-            }
-            $zip_ok = extract_zip_to($location, $smooth_dir);
-            if($zip_ok)
-            {
-                // extract any zips in smooth_dir
-                $zips = glob("$smooth_dir/*.zip");
-                foreach($zips as $zip)
+                $project->delete_smoothreading_dir();
+                $smooth_dir = "$project->dir/smooth";
+                if(!mkdir($smooth_dir))
+                {
+                    throw new FileUploadException("Could not create smooth directory");
+                }
+                $zip_ok = extract_zip_to($location, $smooth_dir);
+                if($zip_ok)
                 {
-                    $zip_ok = extract_zip_to($zip, $smooth_dir);
-                    if(!$zip_ok)
+                    // extract any zips in smooth_dir
+                    $zips = glob("$smooth_dir/*.zip");
+                    foreach($zips as $zip)
                     {
-                        break;
+                        $zip_ok = extract_zip_to($zip, $smooth_dir);
+                        if(!$zip_ok)
+                        {
+                            break;
```

Comment:
Is there a reason we don't just raise the exception here instead of setting $zip_ok and raising later?